### DISCUSSION TITLE
Zustand Store does not load initial objects

### DISCUSSION THREAD

**Discussion by FelixJncs** [neuefische/questions#616] [Web] [answered]
Hey everybody, 

yesterday the Errors were fixed and it did work as expected but today the initial data does only load for a second and then disappears. 

There is no Error message and everything does store like it should. 

This is my Zustand Store: 

```
`import { useState, useEffect } from "react";
import { create } from "zustand";
import { persist } from "zustand/middleware"; //Middelware: intercept state updates and perform actions before or after the update occurs
// Persist: the state of the store will be automatically persisted in localStorage whenever it changes
const useHabitsStore = createLocalStorageStore(
  (set) => ({
    habits: [
      {
        id: "1",
        name: "Procrastination",
        reason: "To avoid or delay completing tasks or responsibilities",
        feeling: "Stressed or overwhelmed",
        overcome:
          "Break tasks into smaller pieces, set deadlines, and use a planner or task list to stay on track.",
        progress: 0,
      },
      {
        id: "2",
        name: "Nail-biting",
        reason: "A nervous or anxious habit",
        feeling: "Embarrassed or ashamed",
        overcome:
          "Use stress-reducing techniques such as exercise, meditation, or deep breathing. Wear gloves or paint your nails to reduce temptation.",
        progress: 0,
      },
      {
        id: "3",
        name: "Smoking",
        reason: "An addiction to nicotine",
        feeling: "Anxious or irritable without nicotine",
        overcome:
          "Talk to a healthcare professional about quitting strategies, such as nicotine replacement therapy or medication. Seek support from friends, family, or a support group.",
        progress: 0,
      },
 ],
    addHabit: (newHabit) =>
      set((state) => ({ habits: [...state.habits, newHabit] })),
    updateHabit: (updatedHabit) =>
      set((state) => ({
        habits: state.habits.map((habit) =>
          habit.id === updatedHabit.id ? updatedHabit : habit
        ),
      })),
    deleteHabit: (habitId) =>
      set((state) => ({
        habits: state.habits.filter((habit) => habit.id !== habitId),
      })),
  }),

  "habits-store" // unique name
);

// A function that returns a custom hook that can be used to access the state.
function createLocalStorageStore(initialStore, name) {
  // Create a Zustand store without persistence.
  const useServerStore = create(initialStore);
  // Create a Zustand store with persistence to local storage.
  const useClientStore = create(persist(initialStore, { name }));

  // A custom hook that selects the appropriate store based on whether
  // the component is hydrated or not.
  function useStore(selector, compare) {
    const [hydrated, setHydrated] = useState(false);

    useEffect(() => {
      setHydrated(true);
    }, []);

    const clientStore = useClientStore(selector, compare);
    const serverStore = useServerStore(selector, compare);

    return hydrated ? clientStore : serverStore;
  }

  return useStore;
}

export default useHabitsStore;
`

```

In this const: `const useServerStore = create(initialStore);` it says the create is deprecated, but i am not sure why. 

[My project](https://github.com/FelixJncs/capstone-project/tree/main)
The Branch with changes made for this Issue is [Branch UserStory7](https://github.com/FelixJncs/capstone-project/pull/16/files) it is already merged because it did work yesterday and was approved. 

i would appreciate some help 😃

**Comment by onemanwenttomow** [comment]
@FelixJncs 

Did you check the value of what's in local storage? 

I think the store first loads the values that you have added for habits (in the store), before checking if there is something to read from in local storage. If there is a key of `habits-store` in local-storage it will replace habits with whatever it finds there (even if it is an empty array). 

You can check what's in local-storage by going to the `application` tab in the dev tools. (chrome)

<img width="1434" alt="Screenshot 2023-05-24 at 14 35 00" src="https://github.com/neuefische/questions/assets/1374487/bb7d32d4-bc15-4285-aebc-bff16d511de8">

**Reply by FelixJncs** [reply]
You are right the initial values do load an are immediately replaced. There is a key habits-store. But now i am not too sure how to fix that. Do i try to load the initial values after the store is checking if there is something in local Storage ?

**Reply by onemanwenttomow** [reply, marked as answer]
I think it depends what you want local storage for.

Do you want to always see those initial habits when the page loads? No matter how the user may have previously interacted with your app. If so then I don't think you want to use local-storage.

Do you want the app to "remember" any changes the user made since their last visit? Then I think it's working as expected at the moment.

**Reply by FelixJncs** [reply]
Thank you for the fast reply, deleting the local Storage and reload was the easy solution 😅 I want the app to "remember" any changes the user made since their last visit so everything works fine now 😁

**Reply by onemanwenttomow** [reply]
No problem. I looked into the issue regarding `create` being deprecated. 

According to the documentation, there is an alternative way to set things up. [See here](https://docs.pmnd.rs/zustand/integrations/persisting-store-data#usage-in-next.js).

I don't think it is necessary to change things on your end, but if you wanted to i also updated my own [zustand demo](https://github.com/onemanwenttomow/zustand-demo/commit/dea0d8874f3e03666d86e260786d667ffc05bc32) as a reference.

**Reply by FelixJncs** [reply]
Ok, thank you 😀 i will look into it. But if its not necessary i will probably let it slide 🙂
